File: java-apps/src/main/java/com/github/stephengold/sportjolt/javaapp/sample/HelloNarrowPhase.java
/*
 Copyright (c) 2020-2025 Stephen Gold and Yanis Boudiaf

 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions are met:

 1. Redistributions of source code must retain the above copyright notice, this
    list of conditions and the following disclaimer.

 2. Redistributions in binary form must reproduce the above copyright notice,
    this list of conditions and the following disclaimer in the documentation
    and/or other materials provided with the distribution.

 3. Neither the name of the copyright holder nor the names of its
    contributors may be used to endorse or promote products derived from
    this software without specific prior written permission.

 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
package com.github.stephengold.sportjolt.javaapp.sample;

import com.github.stephengold.joltjni.AllHitCollideShapeCollector;
import com.github.stephengold.joltjni.Body;
import com.github.stephengold.joltjni.BodyCreationSettings;
import com.github.stephengold.joltjni.BodyInterface;
import com.github.stephengold.joltjni.BroadPhaseLayerFilter;
import com.github.stephengold.joltjni.CapsuleShape;
import com.github.stephengold.joltjni.CharacterRef;
import com.github.stephengold.joltjni.CharacterSettings;
import com.github.stephengold.joltjni.CollideShapeSettings;
import com.github.stephengold.joltjni.ConvexHullShapeSettings;
import com.github.stephengold.joltjni.ObjectLayerFilter;
import com.github.stephengold.joltjni.PhysicsSystem;
import com.github.stephengold.joltjni.Plane;
import com.github.stephengold.joltjni.PlaneShape;
import com.github.stephengold.joltjni.Quat;
import com.github.stephengold.joltjni.RMat44;
import com.github.stephengold.joltjni.RVec3;
import com.github.stephengold.joltjni.SpecifiedObjectLayerFilter;
import com.github.stephengold.joltjni.Vec3;
import com.github.stephengold.joltjni.enumerate.EActivation;
import com.github.stephengold.joltjni.enumerate.EMotionType;
import com.github.stephengold.joltjni.operator.Op;
import com.github.stephengold.joltjni.readonly.ConstNarrowPhaseQuery;
import com.github.stephengold.joltjni.readonly.ConstPlane;
import com.github.stephengold.joltjni.readonly.ConstShape;
import com.github.stephengold.joltjni.readonly.RMat44Arg;
import com.github.stephengold.joltjni.readonly.RVec3Arg;
import com.github.stephengold.joltjni.readonly.Vec3Arg;
import com.github.stephengold.sportjolt.Constants;
import com.github.stephengold.sportjolt.Geometry;
import com.github.stephengold.sportjolt.Mesh;
import com.github.stephengold.sportjolt.TextureKey;
import com.github.stephengold.sportjolt.input.InputProcessor;
import com.github.stephengold.sportjolt.input.RotateMode;
import com.github.stephengold.sportjolt.physics.BasePhysicsApp;
import com.github.stephengold.sportjolt.physics.PhysicsTickListener;
import com.github.stephengold.sportjolt.physics.ShapeSummary;
import org.lwjgl.glfw.GLFW;

/**
 * A simple example of a narrow-phase query.
 * <p>
 * Press the arrow keys to walk. Press the space bar to jump.
 * <p>
 * Builds upon HelloBroadPhase.
 *
 * @author Stephen Gold sgold@sonic.net
 */
final public class HelloNarrowPhase
        extends BasePhysicsApp
        implements PhysicsTickListener {
    // *************************************************************************
    // constants

    /**
     * location of the ghost's center
     */
    final private static Vec3Arg ghostCenter = new Vec3(15f, 0f, -13f);
    // *************************************************************************
    // fields

    /**
     * reusable hit collector
     */
    private static AllHitCollideShapeCollector collector;
    /**
     * true when the space bar is pressed, otherwise false
     */
    private static volatile boolean jumpRequested;
    /**
     * true when the DOWN key is pressed, otherwise false
     */
    private static volatile boolean walkBackward;
    /**
     * true when the UP key is pressed, otherwise false
     */
    private static volatile boolean walkForward;
    /**
     * true when the LEFT key is pressed, otherwise false
     */
    private static volatile boolean walkLeft;
    /**
     * true when the RIGHT key is pressed, otherwise false
     */
    private static volatile boolean walkRight;
    /**
     * broadphase-layer filter that has no effect
     */
    private static BroadPhaseLayerFilter filterBpLayerNoOp;
    /**
     * character to trigger the ghost shape
     */
    private static CharacterRef character;
    /**
     * ghost shape for detecting intrusions
     */
    private static ConstShape ghost;
    /**
     * visualize the ghost shape
     */
    private static Geometry ghostGeometry;
    /**
     * object-layer filter that accepts only moving bodies
     */
    private static ObjectLayerFilter filterObjLayerMoving;
    // *************************************************************************
    // constructors

    /**
     * Instantiate the HelloNarrowphase application.
     * <p>
     * This no-arg constructor was made explicit to avoid javadoc warnings from
     * JDK 18+.
     */
    public HelloNarrowPhase() {
        // do nothing
    }
    // *************************************************************************
    // new methods exposed

    /**
     * Main entry point for the HelloNarrowphase application.
     *
     * @param arguments array of command-line arguments (not null)
     */
    public static void main(String[] arguments) {
        HelloNarrowPhase application = new HelloNarrowPhase();
        application.start();
    }
    // *************************************************************************
    // BasePhysicsApp methods

    /**
     * Create the PhysicsSystem. Invoked once during initialization.
     *
     * @return a new object
     */
    @Override
    protected PhysicsSystem createSystem() {
        // For simplicity, use a single broadphase layer:
        int maxBodies = 2;
        int numBpLayers = 1;
        PhysicsSystem result = createSystem(maxBodies, numBpLayers);

        // To enable the callbacks, register this app as a tick listener:
        addTickListener(this);

        return result;
    }

    /**
     * Initialize the application. Invoked once.
     */
    @Override
    protected void initialize() {
        super.initialize();

        setVsync(true);
        configureCamera();
        configureInput();
        configureLighting();

        // Initialize collector and filters:
        collector = new AllHitCollideShapeCollector();
        filterBpLayerNoOp = new BroadPhaseLayerFilter();
        filterObjLayerMoving = new SpecifiedObjectLayerFilter(objLayerMoving);

        // Create a ghost shape:
        float radius = 10f;
        ConvexHullShapeSettings ss = new ConvexHullShapeSettings(
                new Vec3(radius, 0f, 0f), new Vec3(-radius, 0f, 0f),
                new Vec3(0f, radius, 0f), new Vec3(0f, -radius, 0f),
                new Vec3(0f, 0f, radius), new Vec3(0f, 0f, -radius));
        ghost = ss.create().get();

        // Visualize the ghost shape:
        ShapeSummary summary = new ShapeSummary(ghost, "low/Facet");
        Mesh mesh = meshForShape(ghost, summary);
        ghostGeometry = new Geometry(mesh)
                .setLocation(ghostCenter);
    }

    /**
     * Populate the PhysicsSystem. Invoked once during initialization.
     */
    @Override
    protected void populateSystem() {
        // Create a character with a capsule shape and add it to the system:
        float capsuleRadius = 3f; // meters
        float capsuleHeight = 4f; // meters
        ConstShape shape = new CapsuleShape(capsuleHeight / 2f, capsuleRadius);

        CharacterSettings settings = new CharacterSettings();
        settings.setShape(shape);

        RVec3Arg startLocation = new RVec3(0., 3., 0.);
        long userData = 0L;
        character = new com.github.stephengold.joltjni.Character(
                settings, startLocation, new Quat(), userData, physicsSystem)
                .toRef();
        character.addToPhysicsSystem();

        // Visualize the character:
        visualizeShape(character);

        // Add a plane to represent the ground:
        float groundY = -2f;
        addPlane(groundY);
    }
    // *************************************************************************
    // PhysicsTickListener methods

    /**
     * Callback invoked (by Sport-Jolt, not by Jolt Physics) after the system
     * has been stepped.
     *
     * @param system the system that was just stepped (not null)
     * @param timeStep the duration of the simulation step (in seconds, &ge;0)
     */
    @Override
    public void physicsTick(PhysicsSystem system, float timeStep) {
        // Update the character:
        float maxSeparation = 0.1f; // meters above the ground
        character.postSimulation(maxSeparation);

        // Collect all movable bodies intersecting the ghost shape:
        ConstNarrowPhaseQuery query = physicsSystem.getNarrowPhaseQuery();
        Vec3Arg scale = Vec3.sOne();
        RMat44Arg transform = RMat44.sTranslation(ghostCenter.toRVec3());
        CollideShapeSettings settings = new CollideShapeSettings();
        RVec3Arg base = ghostCenter.toRVec3();
        collector.reset();
        query.collideShape(ghost, scale, transform, settings, base,
                collector, filterBpLayerNoOp, filterObjLayerMoving);

        // Update the color of the ghost:
        int numHits = collector.countHits();
        if (numHits > 0) {
            // Intruder detected!
            ghostGeometry.setColor(Constants.RED);
        } else {
            ghostGeometry.setColor(Constants.YELLOW);
        }
    }

    /**
     * Callback invoked (by Sport-Jolt, not by Jolt Physics) before the system
     * is stepped.
     *
     * @param system the system that's about to be stepped (not null)
     * @param timeStep the duration of the simulation step (in seconds, &ge;0)
     */
    @Override
    public void prePhysicsTick(PhysicsSystem system, float timeStep) {
        Vec3 velocity = character.getLinearVelocity();

        // Clear any horizontal motion from the previous simulation step:
        velocity.setX(0f);
        velocity.setZ(0f);

        // If the character is supported, cause it respond to keyboard input:
        if (character.isSupported()) {
            if (jumpRequested) {
                // Cause the character to jump:
                velocity.setY(18f);

            } else {
                // Walk as directed by the arrow keys:
                Vec3 component1 = cam.getDirection();
                float backward = walkBackward ? 1f : 0f;
                float forward = walkForward ? 1f : 0f;
                component1.scaleInPlace(forward - backward);

                float right = walkRight ? 1f : 0f;
                float left = walkLeft ? 1f : 0f;
                Vec3 component2 = cam.getRight();
                component2.scaleInPlace(right - left);
                Op.assign(velocity, Op.plus(component1, component2));

                velocity.setY(0f);
                if (velocity.length() > 0f) {
                    float scale = 7f / velocity.length();
                    velocity.scaleInPlace(scale);
                }
            }
        }

        character.setLinearVelocity(velocity);
    }
    // *************************************************************************
    // private methods

    /**
     * Add a static horizontal plane body to the system.
     *
     * @param y the desired elevation (in system coordinates)
     */
    private void addPlane(float y) {
        ConstPlane plane = new Plane(0f, 1f, 0f, -y);
        PlaneShape shape = new PlaneShape(plane);
        BodyCreationSettings bcs = new BodyCreationSettings();
        bcs.setMotionType(EMotionType.Static);
        bcs.setObjectLayer(objLayerNonMoving);
        bcs.setShape(shape);

        BodyInterface bi = physicsSystem.getBodyInterface();
        Body body = bi.createBody(bcs);
        bi.addBody(body, EActivation.DontActivate);

        // Visualize the body:
        String resourceName = "/Textures/greenTile.png";
        float maxAniso = 16f;
        TextureKey textureKey
                = new TextureKey("classpath://" + resourceName, maxAniso);
        visualizeShape(body, 0.1f)
                .setSpecularColor(Constants.DARK_GRAY)
                .setTexture(textureKey);
    }

    /**
     * Configure the camera, projection, and CIP during initialization.
     */
    private static void configureCamera() {
        getCameraInputProcessor().setRotationMode(RotateMode.DragLMB);
        cam.setAzimuth(-1.9f)
                .setLocation(35f, 35f, 60f)
                .setUpAngle(-0.5f);
        getProjection().setFovyDegrees(30f);
    }

    /**
     * Configure keyboard input during initialization.
     */
    private void configureInput() {
        getInputManager().add(new InputProcessor() {
            @Override
            public void onKeyboard(int glfwKeyId, boolean isPressed) {
                switch (glfwKeyId) {
                    case GLFW.GLFW_KEY_SPACE:
                        jumpRequested = isPressed;
                        return;

                    case GLFW.GLFW_KEY_DOWN:
                        walkBackward = isPressed;
                        return;
                    case GLFW.GLFW_KEY_LEFT:
                        walkLeft = isPressed;
                        return;
                    case GLFW.GLFW_KEY_RIGHT:
                        walkRight = isPressed;
                        return;
                    case GLFW.GLFW_KEY_UP:
                        walkForward = isPressed;
                        return;

                    default:
                }
                super.onKeyboard(glfwKeyId, isPressed);
            }
        });
    }

    /**
     * Configure lighting and the background color.
     */
    private void configureLighting() {
        setLightDirection(7f, 3f, 5f);

        // Set the background color to light blue:
        setBackgroundColor(Constants.SKY_BLUE);
    }
}

File: jython-apps/scripts/sport/tutorial/hello_narrow_phase.py
"""
 Copyright (c) 2025-2026 Stephen Gold and Yanis Boudiaf

 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions are met:

 1. Redistributions of source code must retain the above copyright notice, this
    list of conditions and the following disclaimer.

 2. Redistributions in binary form must reproduce the above copyright notice,
    this list of conditions and the following disclaimer in the documentation
    and/or other materials provided with the distribution.

 3. Neither the name of the copyright holder nor the names of its
    contributors may be used to endorse or promote products derived from
    this software without specific prior written permission.

 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
"""

# location of the ghost's center
GHOST_CENTER = Vec3(15.0, 0.0, -13.0)


class HelloNarrowPhase(BasePhysicsApp, PhysicsTickListener):
    """
    A simple example of a narrow-phase query.

    Press the arrow keys to walk. Press the space bar to jump.

    Builds upon HelloBroadPhase.

    author:  Stephen Gold sgold@sonic.net
    """

    def createSystem(self):
        "Create the PhysicsSystem. Invoked once during initialization."

        # For simplicity, use a single broadphase layer:
        max_bodies = 2
        num_bp_layers = 1
        result = self.super__createSystem(max_bodies, num_bp_layers)

        # To enable the callbacks, register the application as a tick listener:
        self.addTickListener(self)

        return result

    def initialize(self):
        "Initialize the application. Invoked once."

        self.super__initialize()
        self.setVsync(True)
        self.configure_camera()
        self.configure_input()
        self.configure_lighting()

        # Initialize collector and filters:
        global COLLECTOR
        COLLECTOR = AllHitCollideShapeCollector()

        global FILTER_BP_LAYER_NO_OP
        FILTER_BP_LAYER_NO_OP = BroadPhaseLayerFilter()

        global FILTER_OBJ_LAYER_MOVING
        FILTER_OBJ_LAYER_MOVING = SpecifiedObjectLayerFilter(self.objLayerMoving)

        # Create a ghost shape:
        radius = 10.0
        ss = ConvexHullShapeSettings(
            Vec3(radius, 0.0, 0.0),
            Vec3(-radius, 0.0, 0.0),
            Vec3(0.0, radius, 0.0),
            Vec3(0.0, -radius, 0.0),
            Vec3(0.0, 0.0, radius),
            Vec3(0.0, 0.0, -radius),
        )
        global GHOST
        GHOST = ss.create().get()

        # Visualize the ghost shape:
        summary = ShapeSummary(GHOST, "low/Facet")
        mesh = self.meshForShape(GHOST, summary)
        global GHOST_GEOMETRY
        GHOST_GEOMETRY = Geometry(mesh)
        GHOST_GEOMETRY.setLocation(GHOST_CENTER)

        global JUMP_REQUESTED
        JUMP_REQUESTED = False

        global WALK_BACKWARD
        WALK_BACKWARD = False

        global WALK_FORWARD
        WALK_FORWARD = False

        global WALK_LEFT
        WALK_LEFT = False

        global WALK_RIGHT
        WALK_RIGHT = False

    def populateSystem(self):
        "Populate the PhysicsSystem. Invoked once during initialization."

        # Create a character with a capsule shape and add it to the system:
        capsule_radius = 3.0  # meters
        capsule_height = 4.0  # meters
        shape = CapsuleShape(capsule_height / 2.0, capsule_radius)

        settings = CharacterSettings()
        settings.setShape(shape)

        start_location = RVec3(0.0, 3.0, 0.0)
        user_data = 0
        global CHARACTER
        CHARACTER = Character(
            settings, start_location, Quat(), user_data, self.physicsSystem
        )
        CHARACTER.addToPhysicsSystem()

        # Visualize the character:
        self.visualizeShape(CHARACTER)

        # Add a plane to represent the ground:
        ground_y = -2.0
        self.add_plane(ground_y)

    def physicsTick(self, system, time_step):
        "Callback invoked (by Sport-Jolt, not by Jolt Physics) after the system has been stepped."

        # Update the character:
        max_separation = 0.1  # meters above the floor
        CHARACTER.postSimulation(max_separation)

        # Collect all movable bodies intersecting the ghost shape:
        query = self.physicsSystem.getNarrowPhaseQuery()
        scale = Vec3.sOne()
        transform = RMat44.sTranslation(GHOST_CENTER.toRVec3())
        settings = CollideShapeSettings()
        base = GHOST_CENTER.toRVec3()
        COLLECTOR.reset()
        query.collideShape(
            GHOST,
            scale,
            transform,
            settings,
            base,
            COLLECTOR,
            FILTER_BP_LAYER_NO_OP,
            FILTER_OBJ_LAYER_MOVING,
        )

        # Update the color of the ghost:
        num_hits = COLLECTOR.countHits()
        if num_hits > 0:
            # Intruder detected!
            GHOST_GEOMETRY.setColor(Constants.RED)
        else:
            GHOST_GEOMETRY.setColor(Constants.YELLOW)

    def prePhysicsTick(self, system, time_step):
        "Callback invoked (by Sport-Jolt, not by Jolt Physics) before the system is stepped."

        velocity = CHARACTER.getLinearVelocity()

        # Clear any horizontal motion from the previous simulation step:
        velocity.setX(0.0).setZ(0.0)

        # If the character is supported, make it respond to keyboard input:
        if CHARACTER.isSupported():
            if JUMP_REQUESTED:
                # Cause the character to jump:
                velocity.setY(18.0)

            else:
                # Walk as directed by the arrow keys:
                cam = self.getCamera()
                component1 = cam.getDirection()
                backward = 1.0 if WALK_BACKWARD else 0.0
                forward = 1.0 if WALK_FORWARD else 0.0
                component1.scaleInPlace(forward - backward)

                right = 1.0 if WALK_RIGHT else 0.0
                left = 1.0 if WALK_LEFT else 0.0
                component2 = cam.getRight()
                component2.scaleInPlace(right - left)
                Op.assign(velocity, Op.plus(component1, component2))

                velocity.setY(0.0)
                if velocity.length() > 0.0:
                    scale = 7.0 / velocity.length()
                    velocity.scaleInPlace(scale)

        CHARACTER.setLinearVelocity(velocity)

    def add_plane(self, y):
        "Add a static horizontal plane body to the system."

        plane = Plane(0.0, 1.0, 0.0, -y)
        shape = PlaneShape(plane)
        bcs = BodyCreationSettings()
        bcs.setMotionType(EMotionType.Static)
        bcs.setObjectLayer(self.objLayerNonMoving)
        bcs.setShape(shape)

        bi = self.physicsSystem.getBodyInterface()
        body = bi.createBody(bcs)
        bi.addBody(body, EActivation.DontActivate)

        # Visualize the body:
        resourceName = "/Textures/greenTile.png"
        maxAniso = 16.0
        textureKey = TextureKey("classpath://" + resourceName, maxAniso)
        geometry = self.visualizeShape(body, 0.1)
        geometry.setSpecularColor(Constants.DARK_GRAY)
        geometry.setTexture(textureKey)

    def configure_camera(self):
        "Configure the camera, projection, and CIP during initialization."

        self.getCameraInputProcessor().setRotationMode(RotateMode.DragLMB)

        cam = self.getCamera()
        cam.setAzimuth(-1.9)
        cam.setLocation(35.0, 35.0, 60.0)
        cam.setUpAngle(-0.5)

        self.getProjection().setFovyDegrees(30.0)

    def configure_input(self):
        "Configure keyboard input during initialization."

        processor = AnonymousInputProcessor()
        self.getInputManager().add(processor)

    def configure_lighting(self):
        "Configure lighting and the background color."

        self.setLightDirection(7.0, 3.0, 5.0)

        # Set the background color to light blue:
        self.setBackgroundColor(Constants.SKY_BLUE)


class AnonymousInputProcessor(InputProcessor):
    "When the SPACE key is pressed, jump.  When the W key is pressed, walk forward."

    def onKeyboard(self, glfw_key_id, is_pressed):
        if glfw_key_id == GLFW.GLFW_KEY_SPACE:
            global JUMP_REQUESTED
            JUMP_REQUESTED = is_pressed
            return

        elif glfw_key_id == GLFW.GLFW_KEY_DOWN:
            global WALK_BACKWARD
            WALK_BACKWARD = is_pressed
            return

        elif glfw_key_id == GLFW.GLFW_KEY_LEFT:
            global WALK_LEFT
            WALK_LEFT = is_pressed
            return

        elif glfw_key_id == GLFW.GLFW_KEY_RIGHT:
            global WALK_RIGHT
            WALK_RIGHT = is_pressed
            return

        elif glfw_key_id == GLFW.GLFW_KEY_UP:
            global WALK_FORWARD
            WALK_FORWARD = is_pressed
            return

        self.super__onKeyboard(glfw_key_id, is_pressed)


application = HelloNarrowPhase()
application.start("HelloNarrowPhase")
